subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java

```diff
@@ -115,24 +120,31 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
 
         boolean buildCacheEnabled = buildCacheController.isEnabled();
         boolean scanPluginApplied = buildScanPlugin.isBuildScanPluginApplied();
+        TaskCacheKeyCalculator cacheKeyCalculator = new DefaultTaskCacheKeyCalculator();
 
         TaskExecuter executer = new ExecuteActionsTaskExecuter(
             buildCacheEnabled,
+            fingerprinterRegistry,
+            executionHistoryStore,
+            outputFilesRepository,
             buildOperationExecutor,
             asyncWorkTracker,
             actionListener,
             workExecutor
         );
+        executer = new ResolveIncrementalChangesTaskExecuter(executer);
         executer = new ResolveTaskOutputCachingStateExecuter(buildCacheEnabled, relativeFilePathResolver, executer);
         if (buildCacheEnabled || scanPluginApplied) {
-            executer = new ResolveBuildCacheKeyExecuter(executer, buildOperationExecutor, buildCacheController.isEmitDebugLogging());
+            executer = new ResolveBuildCacheKeyExecuter(buildOperationExecutor, cacheKeyCalculator, buildCacheController.isEmitDebugLogging(), executer);
         }
+        executer = new ResolveBeforeExecutionStateTaskExecuter(classLoaderHierarchyHasher, valueSnapshotter, fingerprinterRegistry, executer);
         executer = new ValidatingTaskExecuter(executer);
```

Comment:
`ValidatingTaskExecuter` could re-use the input snapshots from `ResolveBeforeExecutionStateTaskExecuter` instead of checking manually for existence of files and types, right?

Reply:
Sounds right, would you mind adding a card to do this?

Reply:
https://github.com/gradle/build-cache/issues/1389

Reply:
Thanks!